File: leetcode/queue_stack/2534_Time_Taken_to_Cross_the_Door/codes/queue.c
#include <stdlib.h>

int* timeTaken(int* arrival, int arrivalSize, int* state, int stateSize, int* returnSize) {
    int n = arrivalSize;
    int* res = (int*)malloc(n * sizeof(int));
    int* enterQ = (int*)malloc(n * sizeof(int));
    int* exitQ = (int*)malloc(n * sizeof(int));
    
    int enFront = 0, enRear = 0;
    int exFront = 0, exRear = 0;

    int time = 0, i = 0, lastUsed = -1;

    for (int j = 0; j < n; ++j) {
        res[j] = -1;
    }

    while (i < n || enFront < enRear || exFront < exRear) {
        while (i < n && arrival[i] <= time) {
            if (state[i] == 0) {
                enterQ[enRear++] = i;
            } else {
                exitQ[exRear++] = i;
            }
            i++;
        }

        if (exFront < exRear && (lastUsed == 1 || lastUsed == -1 || enFront == enRear)) {
            int person = exitQ[exFront++];
            res[person] = time;
            lastUsed = 1;
        } else if (enFront < enRear) {
            int person = enterQ[enFront++];
            res[person] = time;
            lastUsed = 0;
        } else {
            lastUsed = -1; 
        }

        time++;
    }

    free(enterQ);
    free(exitQ);
    *returnSize = n;
    return res;
}

File: leetcode/queue_stack/2534_Time_Taken_to_Cross_the_Door/codes/queue.py
from collections import deque
from typing import List

class Solution:
    def timeTaken(self, arrival: List[int], state: List[int]) -> List[int]:
        n = len(arrival)
        res = [-1 for _ in range(n)]
        enter_queue = deque()
        exit_queue = deque()

        time = 0
        i = 0  
        last_used = -1  

        while i < n or enter_queue or exit_queue:
            while i < n and arrival[i] <= time:
                if state[i] == 0:
                    enter_queue.append(i)
                else:
                    exit_queue.append(i)
                i += 1

            if exit_queue and (last_used == 1 or last_used == -1 or not enter_queue):
                person = exit_queue.popleft()
                res[person] = time
                last_used = 1
            elif enter_queue:
                person = enter_queue.popleft()
                res[person] = time
                last_used = 0
            else:
                last_used = -1  

            time += 1

        return res
